lib/class.NGIpublications.php

```diff
@@ -243,162 +308,242 @@ public function scorePublication($publication_id) {
+	public function reservePublications($user_email,$year,$score=5,$limit=10) {
+		if($user_email=filter_var($user_email,FILTER_VALIDATE_EMAIL)) {
+			$year=filter_var($year,FILTER_VALIDATE_INT);
+			$score=filter_var($score,FILTER_VALIDATE_INT);
+			$limit=filter_var($limit,FILTER_VALIDATE_INT);
+			if($year && $score && $limit) {
+				// Check if user has already reserved papers
+				if(!$check=sql_fetch("SELECT * FROM publications WHERE reservation_user='$user_email' AND status IS NULL")) {
+					// Only reserve new ones if the old list is empty
+					$timestamp=time();
+					$reserve=sql_query("UPDATE publications 
+						SET 
+							reservation_user='$user_email', 
+							reservation_timestamp='$timestamp' 
+						WHERE 
+							pubdate>='$year-01-01' AND 
+							pubdate<='$year-12-31' AND 
+							score>='$score' AND 
+							(status IS NULL OR status='maybe') AND 
+							reservation_user IS NULL 
+						ORDER BY RAND() LIMIT $limit");
+					
+					// Update log on the reserved papers
+					if($updated=sql_query("SELECT * FROM publications WHERE reservation_user='$user_email' AND reservation_timestamp=$timestamp")) {
+						while($publication=$updated->fetch_assoc()) {
+							$log=$this->addLog('Publication reserved for validation by: '.$user_email,'update',$publication['log']);
+							$update_log=sql_query("UPDATE publications SET log='$log' WHERE id=".$publication['id']);
+						}
+					}
+				}
+				
+				// Fetch all reserved un-verified and 'maybe' papers
+				if($query=sql_query("SELECT * FROM publications WHERE reservation_user='$user_email' AND (status IS NULL OR status='maybe')")) {
+					return $query;
+				} else {
+					return FALSE;
+				}
 			} else {
-				$output='ERROR: could not retrieve publication data';
+				return FALSE;
 			}
 		} else {
-			$output='ERROR: invalid publication ID';
+			return FALSE;
 		}
-		
-		return $output;
 	}
 	
-	
-	public function showSelectedPublications($sql) {
-		if(is_object($sql)) {
-			while($publication=$sql->fetch_assoc()) {
-				$output.=$this->formatPublication($this->publicationData($publication));
+	// Summarize verified publications
+	public function getScoreboard($year=FALSE,$user=FALSE) {
+		$result=array();
+		if($user=filter_var($user,FILTER_VALIDATE_EMAIL)) {
+			if($year=filter_var($year,FILTER_VALIDATE_INT)) {
+				// Get user score for specified year
+				$query=sql_query("SELECT status,COUNT(*) AS count FROM publications 
+					WHERE 
+						(status='verified' OR status='discarded') AND 
+						pubdate>='$year-01-01' AND 
+						pubdate<='$year-12-31' AND 
+						reservation_user='$user' 
+					GROUP BY status 
+					ORDER BY status DESC");
+			} else {
+				// Get total user score
+				$query=sql_query("SELECT status,COUNT(*) AS count FROM publications 
+					WHERE 
+						(status='verified' OR status='discarded') AND 
+						reservation_user='$user' 
+					GROUP BY status 
+					ORDER BY status DESC");
```

Comment:
Oh cool! You implemented a scoreboard. I didn't think we had time for it but you are a PHP guru. I only begrudgingly code PHP :wink:

Reply:
Yes I thought that might give us a little extra incentive to rate some extra papers, only limited testing though... will be easier to test when it's live on databases